WIP: Initial addition of glossary (#222)
JuliaDiff/ChainRulesCore.jl: docs/src/glossary.md

# ChainRules Glossary

This glossary serves as a quick reference for common terms used in the field of Automatic Differentiation, as well as those used throughout the documentation relating specifically to ChainRules.

##Definitions:

###Adjoint:

The conjugate transpose of the Jacobian for a given function `f`.

###Derivative:

> The derivative of a function `y = f(x)` with respect to the independent variable `x` denoted `f'(x)` or `dy/dx` is the rate of change of the dependent variable `y` with respect to the change of the independent variable `x`. In multiple dimensions, we may refer to the gradient of a function.

Reviewer: Correct me if I am wrong, but should
```
the rate of change of the dependent variable `y` with respect to the change of the independent variable `x`.
```

be simply changed to
```
the rate of change of the dependent variable `y` with respect to the independent variable `x`.
```

Author: Maybe something like this is more appropriate:

"The derivative of a function `y = f(x)` with respect to the independent variable `x` denoted `f'(x)` or `dy/dx` is the measure of the rate of change of the dependent variable `y` with respect to the rate of change of the independent variable `x`. In multiple dimensions, we may refer to the gradient of a function."

I agree that it is normally referred to "with respect to x", however I think it is important to explain what is being measured. Would the changes above seem suitable?

Reviewer 2: >  In multiple dimensions, we may refer to the gradient of a function."

Maybe better to say:
in multiple dimensions the derivative is not defined.
Instead there are the partial derivatives, the directional derivative, the jacobian (called gradient for scalar-valued functions).

Author: Updated

# ChainRules Glossary

This glossary serves as a quick reference for common terms used in the field of Automatic Differentiation, as well as those used throughout the documentation relating specifically to ChainRules.

##Definitions:

###Adjoint:

> The conjugate transpose of the Jacobian for a given function `f`.

Reviewer 2: Probably we want to title the Adjoint of a function.
ANd start by saying that the adjoint of a matrix is another word for it's conjugate transpose.
Then mentioning that it can also be applied to a linear operator as every linear operator can be described as `y = Jx` and that as an adjoint linear operator of `y' = x'J'`.
Then say that people say as a shorthand/abuse of terminoly the adjoint of a function,
when what they actually mean is to get a function which is the adjoint of pushfoward linear operator.
The pushforward linear operator is the the linear operator that has the same jacobian as the function at that point.
The pullback is its adjoint.
 linearization of the function at a point, to get a linear operator (the `pushforward`),
and then 
Then say that people occationally say the adjoint of a function,
when what they really mean is: the adjoint of the jacobian of the function,
or they mean the pullback.
Sometimes people say adjoint of a function to mean pullback.

> The conjugate transpose of the Jacobian for a given function `f`.

Author: The adjoint is super broad as a term, so I am having a hard time figuring out how much / how little to include

###Automatic Differentiation:

Automatic Differentiation is the process of applying numerical methods to solving derivative problems, most often algorithmically, using computer programs to achieve high degrees of accuracy.

###Derivative:

The derivative of a function `y = f(x)` with respect to the independent variable `x` denoted `f'(x)` or `dy/dx` is the rate of change of the dependent variable `y` with respect to the change of the independent variable `x`. In multiple dimensions, we may refer to the gradient of a function.

###Differential:

The differential of a given function `y = f(x)` denoted `dy` is the product of the derivative function `f'(x)` and the increment of the independent variable `dx`. In multiple dimensions, it is the sum of these products across each dimension (using the partial derivative and the given independent variable's increment).

> In ChainRules, differentials are types ("differential types") and correspond to primal types. A differential should represent a difference between two primal values.

Reviewer 2: ```suggestion
In ChainRules, differentials are types ("differential types") and correspond to primal types. A differential type should represent a difference between two primal typed values.
```

Author: 👍🏻

####Natural Differential:

> A natural differential type for a given primal type is the type people would intuitively associate with representing the difference between two values of the primal type.

Reviewer 2: Worth also commenting that this is incontrast to the Structal DIfferential type

> A natural differential type for a given primal type is the type people would intuitively associate with representing the difference between two values of the primal type.

Author: Updated

> 

Reviewer 2: ```suggestion
Not to be confused with that [natural gradient](https://towardsdatascience.com/natural-gradient-ce454b3dcdfa), which is an unrelated concept.

```

Author: Added, good call.

####Structural Differential:

> If a given primal type `P` does not have a natural differential, we need to come up with one that makes sense. These are called structural differentials and are represented as `Composite{P, <:NamedTuple}`.

Reviewer 2: ```suggestion
If a given primal type `P` does not have a natural differential, we need to come up with one that makes sense. These are called structural differentials and are represented as `Composite{P}`, and mirror the structure of the primal type.
```

Reviewer 2: again worth mentioning this is a chain rules specific termology

Author: Added

####Semi-Structural Differential:

> A structural differential that contains at least one natural differential field.

Reviewer 2: I would delete this one, its not going to come up outside the page in the docs that introduces that word.
And we are not sold on it in the first place

Author: Okay, good to know! It is gonzo

####Thunk:

> 

Reviewer 2: Thunk is actually more generally a term from computing.

Probably include a standard definition:
https://en.wikipedia.org/wiki/Thunk

In general  a way to represent thunks in julia is as anonoymous functions (including closures) which take no inputs.
e.g. `()->foo(x,y,z)`
And in ChainRules the specific `AbstractThunk` subtype is such a function that is setup to act like a differential.

Author: Added

> An "unnatural" differential type. If we wish to delay the computation of a derivative for whatever reason, we wrap it in a `Thunk` or `ImplaceableThunk`. It holds off on computing the wrapped derivative until it is needed.

Author: It very much should! Good catch :)

####Zero:

> `Zero()` can also be a differential type. If you have trouble understanding the rules enforced upon differential types, consider this one first, as `Zero()` is the trivial vector space.

Reviewer 2: this sentence is unclear

Author: Updated with more fundamental info

###Directional Derivative:

> The directional derivative of a function `f` at any given point in any given unit-direction is the gradient multiplied by the direction. It represents the rate of change of `f` in that direction.

Reviewer 2: mention that the pushforward computes the directional derivative.

Author: Included

> ###F-rule:

Reviewer 2: ```suggestion
### `frule`:

Forward-mode rule.
```

> ###F-rule:
> 
> A function used in forward-mode differentiation. For a given function `f`, it takes in the positional and keyword arguments of `f` and returns the primal result and the pushforward.

Reviewer 2: ```suggestion
### `frule`

A forward mode rule, that descripes how to propagate the sensitivity into the forwards direction.
The `frule` fuses the primal computation and the pushforward, it thus takes in the primal function name, the primal arguments and their matching partial derivatives, and returns the primal output, and the matching directional derviative.
```

Author: This is definitely a more explicit and explanatory def 👍🏻

###Gradient:

> The gradient of a scalar function `f` represented by `∇f` is a vector function whose components are the partial derivatives of `f` with respect to each dimension of the domain of `f`.

Reviewer 2: Mention that this is a special case of the jacobian for scalar values functions (if one is happy considering `$\mathbb{R}$` as being the same as `$\mathbb{R}^1$

Author: Added, without the note about R being considered equivalent to R^1. Is that worth including? If so what is convention for including LaTeX style code in MarkDown?

###Jacobian:

The Jacobian of a vector-valued function `f` is the matrix of `f`'s first-order partial derivatives.

> ###Jacobian Transpose Vector Product (j'vp):

Author: Good call, I moved these under their corresponding super-definitions

The product of the adjoint of the Jacobian and the vector in question. A description of the pullback in terms of its Jacobian.

> ###Jacobian Vector Product (jvp):

Author: added as mentioned above

> The product of the Jacobian and the vector in question. It is a description of the pushforward in terms of its Jacobian.

Reviewer 2: Mention that the term pushforward and jvp are often used interchangably.
Strictly speaking one could say that the pushforward computes the jvp,
but its not usual to see the jvp as the name of the function.

Author: 👍🏻

###Primal:

> Something relating to the original problem, as opposed to relating to the derivative. In ChainRules, primals are types ("primal types").

Reviewer 2: ```suggestion
Something relating to the original problem, as opposed to relating to the derivative.
Such as:
 - The primal function being the function being differnetiated
 - The primal inputs being the inputs to that function (the point that the derivative is being calculated at)
 - The primal outputs being the result of applying the primal function to the primal inputs
 - The primal pass (also called the forward pass) where the computation is run to get the primal outputs (generally before doing a derivative (i.e. reverse pass) in reverse mode AD).
  - The primal computation which is the part of the code that is run during the primal pass and must at least compute the primal outputs (but may compute other things to use during the derivative pass).
 
 - The primal types being the types of the primal inputs/outputs
```

Author: Added these sub-defs as points

###Pullback:

> `Pullback(f)` describes the sensitivity of the input of `f` as a function of (for the relative change to) the sensitivity of the output of `f`. Can be represented as the dot product of a vector (left) the adjoint Jacobian (right).

Reviewer 3: I'm finding the definitions of pullback and pushforward in terms of sensitivity a little confusing.  Would the following still be accurate?

```suggestion
`Pullback(f)` describes the sensitivity of a quantity to the input of `f` as a function of its sensitivity to the output of `f`. Can be represented as the dot product of a vector and the adjoint of the Jacobian.
```

Author: These changes make sense to me. Applying both

###Pushforward:

> `Pushforward(f)` describes the sensitivity of the output of `f` as a function of (for the relative change to) the sensitivity of the input of `f`. Can be represented as the dot product of the Jacobian (left) and a vector (right).

Reviewer 3: See comment on pullback definition.

```suggestion
`Pushforward(f)` describes the sensitivity of a quantity to the output of `f` as a function of its sensitivity to the input of `f`. Can be represented as the dot product of the Jacobian and a vector.
```

Author: Applied as mentioned above

> ###R-rule:
> 

Reviewer 2: ```suggestion
### `rrule`

Reverse mode rules.
```

Author: 👍🏻
